rfcs/0001-messages.md

```diff
@@ -0,0 +1,408 @@
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    SharePackage,
+    SigningCommitments,
+    SigningPackage,
+    SignatureShare,
+    AggregateSignature,
+}
+
+/// The numeric values used to identify the protocol version during serialization.
+struct MsgVersion(u8);
+
+const BASIC_FROST_SERIALIZATION: MsgVersion = MsgVersion(0);
+
+/// The numeric values used to identify each participant during serialization.
+///
+/// In the `frost` module, participant ID `0` should be invalid.
+/// But in serialization, we want participants to be indexed from `0..n`,
+/// where `n` is the number of participants.
+/// This helps us look up their shares and commitments in serialized arrays.
+/// So in serialization, we assign the dealer and aggregator the highest IDs,
+/// and mark those IDs as invalid for signers.
+///
+/// "When performing Shamir secret sharing, a polynomial `f(x)` is used to generate
+/// each party’s share of the secret. The actual secret is `f(0)` and the party with
+/// ID `i` will be given a share with value `f(i)`.
+/// Since a DKG may be implemented in the future, we recommend that the ID `0` be declared invalid."
+/// https://raw.githubusercontent.com/ZcashFoundation/redjubjub/main/zcash-frost-audit-report-20210323.pdf#d
+enum ParticipantId {
+    /// A serialized participant ID for a signer.
+    ///
+    /// Must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    Signer(u8),
+    /// The fixed participant ID for the dealer.
+    Dealer,
+    /// The fixed participant ID for the aggregator.
+    Aggregator,
+}
+
+/// The fixed participant ID for the dealer.
+const DEALER_PARTICIPANT_ID: u8 = u8::MAX - 1;
+
+/// The fixed participant ID for the aggregator.
+const AGGREGATOR_PARTICIPANT_ID: u8 = u8::MAX;
+
+/// The maximum `ParticipantId::Signer` in this serialization format.
+///
+/// We reserve two participant IDs for the dealer and aggregator.
+const MAX_SIGNER_PARTICIPANT_ID: u8 = u8::MAX - 2;
+```
+
+### Payloads
+
+Each payload defines a new message:
+
+```rust
+/// The data required to serialize `frost::SharePackage`.
+///
+/// The dealer sends this message to each signer for this round.
+/// With this, the signer should be able to build a `SharePackage` and use
+/// the `sign()` function.
+///
+/// Note: `frost::SharePackage.public` can be calculated from `secret_share`.
+struct messages::SharePackage {
+    /// This participant's secret key share: `frost::SharePackage.share.value`.
+    secret_share: frost::Scalar,
+    /// Commitment for the signer as a single jubjub::AffinePoint.
+    /// A set of commitments to the coefficients (which themselves are scalars)
+    /// for a secret polynomial _f_: `frost::SharePackage.share.commitment`
+    share_commitment: Vec<jubjub::AffinePoint>,
+    /// The public signing key that represents the entire group:
+    /// `frost::SharePackage.group_public`.
+    group_public: jubjub::AffinePoint,
```

Comment:
I think we need to change the type of `group_public` to `VerificationKey<SpendAuth>` throughout this design.

```suggestion
    /// The public signing key that represents the entire group:
    /// `frost::SharePackage.group_public`.
    group_public: jubjub::VerificationKey<SpendAuth>,
```

Reply:
Or we should explain how to convert from `VerificationKey<SpendAuth>` to `AffinePoint` (or `Scalar`) and back in the serialization section.

We explain how to convert to bytes here: https://github.com/ZcashFoundation/redjubjub/pull/85/files#diff-eb4c7391aa1f02fd5c5b3a11c199bb7a6efe129fb024fc3f31ff1afbd04ee160R306
But that's not the same as converting to a point or scalar.

Reply:
We were looking into that in this comment https://github.com/ZcashFoundation/redjubjub/pull/85#discussion_r629423342.

I think the correct way would be to use [u8; 32] instead of the `AffinePoint` because, as already mentioned, the `VerificationKey` can be built from [u8; 32] by this trait https://github.com/ZcashFoundation/redjubjub/blob/4cec6f87b4e2a4ea58bc3353136c11aed1cba5c7/src/verification_key.rs#L114.

Reply:
Based in this 2 comments it seems this need to be `VerificationKey<SpendAuth>` in the rust side and `[u8; 32]` in the serialization side.

Reply:
I made changes for this at https://github.com/ZcashFoundation/redjubjub/pull/85/commits/1f5f04f9c8107d0754124bf9c0c0fd31c1c5d605